### PR TITLE
Remove TSDX from build system

### LINKED ISSUE #1061: Replace tsdx use in build of the FDC3 NPM module
### Area of Issue
[x] API

~Upgrading tsdx to 0.13.3~ Replacing tsdx (used to bundle the FDC3 NPM module) would resolve 17 moderate vulnerabilities in the FDC3 repo and should ease future maintenance as tsdx is now unmaintained ([ref](https://github.com/jaredpalmer/tsdx/issues/1065)).

### npm audit report
```
jsdom  <=16.5.3
Severity: moderate
Insufficient Granularity of Access Control in JSDom - https://github.com/advisories/GHSA-f4c9-cqv8-9v98
Depends on vulnerable versions of request
Depends on vulnerable versions of request-promise-native
Depends on vulnerable versions of tough-cookie
fix available via `npm audit fix --force`
Will install tsdx@0.13.3, which is a breaking change
node_modules/jsdom
  jest-environment-jsdom  10.0.2 - 25.5.0
  Depends on vulnerable versions of jsdom
  node_modules/jest-environment-jsdom
    jest-config  12.1.1-alpha.2935e14d - 25.5.4
    Depends on vulnerable versions of @jest/test-sequencer
    Depends on vulnerable versions of jest-environment-jsdom
    Depends on vulnerable versions of jest-jasmine2
    node_modules/jest-config
      jest-cli  12.1.1-alpha.2935e14d || 12.1.2-alpha.6230044c - 25.5.4
      Depends on vulnerable versions of @jest/core
      Depends on vulnerable versions of jest-config
      node_modules/jest-cli
        jest  12.1.2-alpha.6230044c - 25.5.4
        Depends on vulnerable versions of @jest/core
        Depends on vulnerable versions of jest-cli
        node_modules/jest
          tsdx  >=0.14.0
          Depends on vulnerable versions of jest
          node_modules/tsdx
      jest-runner  21.0.0-alpha.1 - 25.5.4
      Depends on vulnerable versions of jest-config
      Depends on vulnerable versions of jest-jasmine2
      Depends on vulnerable versions of jest-runtime
      node_modules/jest-runner
        @jest/test-sequencer  <=25.5.4
        Depends on vulnerable versions of jest-runner
        Depends on vulnerable versions of jest-runtime
        node_modules/@jest/test-sequencer
      jest-runtime  12.1.1-alpha.2935e14d - 25.5.4
      Depends on vulnerable versions of jest-config
      node_modules/jest-runtime
        jest-jasmine2  24.2.0-alpha.0 - 25.5.4
        Depends on vulnerable versions of jest-runtime
        node_modules/jest-jasmine2

node-notifier  <8.0.1
Severity: moderate
OS Command Injection in node-notifier - https://github.com/advisories/GHSA-5fw9-fq32-wv5p
fix available via `npm audit fix --force`
Will install tsdx@0.13.3, which is a breaking change
node_modules/node-notifier
  @jest/reporters  <=26.4.0
  Depends on vulnerable versions of node-notifier
  node_modules/@jest/reporters
    @jest/core  <=25.5.4
    Depends on vulnerable versions of @jest/reporters
    Depends on vulnerable versions of jest-config
    Depends on vulnerable versions of jest-runner
    Depends on vulnerable versions of jest-runtime
    node_modules/@jest/core

request  *
Severity: moderate
Server-Side Request Forgery in Request - https://github.com/advisories/GHSA-p8p7-x288-28g6
Depends on vulnerable versions of tough-cookie
fix available via `npm audit fix --force`
Will install tsdx@0.13.3, which is a breaking change
node_modules/request
  request-promise-core  *
  Depends on vulnerable versions of request
  node_modules/request-promise-core
    request-promise-native  >=1.0.0
    Depends on vulnerable versions of request
    Depends on vulnerable versions of request-promise-core
    Depends on vulnerable versions of tough-cookie
    node_modules/request-promise-native

tough-cookie  <4.1.3
Severity: moderate
tough-cookie Prototype Pollution vulnerability - https://github.com/advisories/GHSA-72xf-g2v4-qvf3
fix available via `npm audit fix --force`
Will install tsdx@0.13.3, which is a breaking change
node_modules/request-promise-native/node_modules/tough-cookie
node_modules/request/node_modules/tough-cookie
node_modules/tough-cookie

17 moderate severity vulnerabilities
```

### PR DESCRIPTION
**Pull request by julianna-ciq**
resolves #1061 

TSDX is a build system to uses rollup, prettier, eslint, and jest. It's helpful to use TSDX, because it will manage your build, lint, and testing procedures for you, instead of you having to manage all of those dependencies.

Unfortunately, TSDX hasn't been keeping those dependencies up-to-date, including the dependencies with vulnerabilities. If you have already cloned this repo, go to the `main` branch and run `npm audit`. You'll see 17 moderate vulnerabilities. Those vulnerabilities are all downstream of TSDX, and many of them are for features not even used in this repo. Furthermore, since TSDX relies on so many stale dependencies, it's reasonable to expect that problems will only continue to grow .

To overcome the risk presented by TSDX's stale dependencies, I've replaced TSDX with the rollup, prettier/eslint, and jest dependencies that it was obfuscating. Note that I copied over most of TSDX's *configs* for these tools, to maintain continuity.

The changes in this PR include:
* Linting: Added ESLint, plus plugins, and configured it to run in place of `tsdx lint`
* Testing: Added Jest, plus plugins, and configured it to run in place of `tsdx test`
* Building: Added Rollup, plus plugins, and configured it to run in place of `tsdx build` and `tsdx watch`
* Tidied up some of the scripts in package.json
* Added a `clean` script, which just removes the generated `dist` folder. This was convenient while I was rerunning builds to confirm that I was generating things correctly. Now, `clean` runs before every `build` to avoid the potential of hiding build errors.
* Linted some files. TSDX was using an older version of ESLint recommendations, so some files had to be re-linted to update to the latest recommendations. (Specifically, prefer-const)
* Updated a test. TSDX was using an older version of jest/jsdom. When I updated to the latest jest, I had to enable legacy timer mocks in one of the testing files.

### CODE REVIEW

**Review by kriswest — APPROVED** [review]
Fixed clean on windows and build on linux - module needs testing in the field, but otherwise LGTM. Many thanks for taking this on @julianna-ciq

**Comment by bingenito** [comment]
Tested all changed commands on windows (command prompt) and RHEL9 and look good.  Also very happy to see npm install complete with 0 vulnerabilities.

**Comment by kriswest** [comment]
We only had a couple of minor types conflicts in finsemble when we dropped the new module in - one relates to a change that should only go out in 2.2.0 that's already merged and the second maybe a type issue caused by that first one. Hence, I'll figure out what to do to back out the change temporarily so we can get this released!

**Comment by kriswest** [comment]
The change that needs backing out is: 
- https://github.com/finos/FDC3/pull/1170

File: package.json [hunk 1]
```diff
@@ -42,15 +41,33 @@
...
     "quicktype": "23.0.78",
-    "tsdx": "^0.14.1",
+    "rimraf": "^5.0.5",
+    "rollup": "4.12.1",
+    "ts-jest": "29.1.2",
     "tslib": "^2.0.1",
     "typescript": "^4.0.3"
+  },
+  "optionalDependencies": {
+    "@rollup/rollup-linux-x64-gnu": "4.14.1"
```
[the remark this answers is not in the document]
Comment (kriswest) [reply package.json:71]: I added this after it started blowing up on netlify/github actions. Doesn't seem to need the window version declared though (just works there). I haven't tested on a mac - all you need to do to check is `npm run build`.

**Comment by bingenito** [comment]
Did it always build upon npm install?  I now get eslint failures that I wasn't seeing on last test -

```
> @finos/fdc3@2.1.0-beta.8 lint
> eslint src/ test/ --ext .ts --fix

C:\Users\brian\Source\FDC3\src\context\ContextType.ts
  46:85  error  Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `object` instead.
- If you want a type meaning "any value", you probably want `unknown` instead.
- If you want a type meaning "empty object", you probably want `Record<string, never>` instead.
- If you really want a type meaning "any non-nullish value", you probably want `NonNullable<unknown>` instead  @typescript-eslint/ban-types

C:\Users\brian\Source\FDC3\src\intents\Intents.ts
  32:49  error  Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `object` instead.
- If you want a type meaning "any value", you probably want `unknown` instead.
- If you want a type meaning "empty object", you probably want `Record<string, never>` instead.
- If you really want a type meaning "any non-nullish value", you probably want `NonNullable<unknown>` instead  @typescript-eslint/ban-types

✖ 2 problems (2 errors, 0 warnings)
```

**Comment by kriswest** [comment]
It did indeed always build on install, not sure why. 

These two new eslint errors are from: 
- #1139

We'll should adjust as it suggests to clear them out...

**Comment by kriswest** [comment]
@bingenito @andreifloricel it took me a little while to understand what was going on here (https://github.com/InteropIO/FDC3/blob/c2cb9d51e8644ac64e3640b1eb0cb95bb9cc8d5b/src/context/ContextType.ts#L46-L48). 

It would be nice if, when using the type, if Intellisense or similar was proposing the standard types for you - although you can create your own so any string is allowed:

```ts
export type ContextType = StandardContextType | ExperimentalContextType | string;
```

However, the typescript compiler aggressively reduces that to just `string`, which is correct for type safety, but useless for autocomplete. @andreifloricel clearly knows this and used the standard workaround

```ts
export type ContextType = StandardContextType | ExperimentalContextType | (string & {});
```

However, that trips the modern linting rule `@typescript-eslint/ban-types`. I couldn't find another composition that achieves the same, so I've just disabled that linting rule for the two lines concerned, and all is well again.

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (bingenito) [reply package.json:71]: > I added this after it started blowing up on netlify/github actions. Doesn't seem to need the window version declared though (just works there). I haven't tested on a mac - all you need to do to check is `npm run build`.

Just tested mac again after linting issue was fixed and it is all good.

**Review by bingenito — APPROVED** [review]

**Pull request merged**